kayak/twopc_runner.go

```diff
@@ -0,0 +1,785 @@
+		}
+
+		// commit log
+		r.stableStore.SetUint64(keyCommittedIndex, index)
+		r.lastLogHash = &lastLog.Hash
+		r.lastLogIndex = lastLog.Index
+		r.lastLogTerm = lastLog.Term
+
+		// set state to idle
+		r.setState(Idle)
+
+		return nil
+	}))
+}
+
+func (r *TwoPCRunner) processRollback(req Request) {
+	// rollback log
+	req.SendResponse(nestedTimeoutCtx(context.Background(), r.config.RollbackTimeout, func(ctx context.Context) (err error) {
+		// TODO(xq262144), check current running transaction index
+		if r.currentState != Prepared {
+			// not prepared, failed directly
+			return ErrInvalidRequest
+		}
+
+		// get index
+		var index uint64
+		if index, err = r.decodeLogIndex(req.GetRequest()); err != nil {
+			return
+		}
+
+		var lastIndex uint64
+		if lastIndex, err = r.logStore.LastIndex(); err != nil {
+			return err
+		} else if lastIndex < index {
+			// not logged, no rollback required, maybe previous initiated rollback
+			return nil
+		}
+
+		if r.lastLogIndex+1 != index {
+			// not at the head of the commit position
+			return ErrInvalidLog
+		}
+
+		// get log
+		var lastLog Log
+		if err = r.logStore.GetLog(index, &lastLog); err != nil {
+			return err
+		}
+
+		// decode log
+		var decodedLog interface{}
+		if decodedLog, err = r.decodeLogData(lastLog.Data); err != nil {
+			return err
+		}
+
+		// rollback on storage
+		if err = r.config.Storage.Rollback(ctx, decodedLog); err != nil {
+			return err
+		}
+
+		// rewind log, can be failed, since committedIndex is not updated
+		r.logStore.DeleteRange(r.lastLogIndex+1, index)
+
+		// set state to idle
+		r.setState(Idle)
+
+		return nil
+	}))
+}
+
+// Start a goroutine and properly handle the race between a routine
+// starting and incrementing, and exiting and decrementing.
+func (r *TwoPCRunner) goFunc(f func()) {
+	r.routinesGroup.Add(1)
+	go func() {
+		defer r.routinesGroup.Done()
+		f()
+	}()
+}
+
+// NewTwoPCWorkerWrapper returns a wrapper for remote worker
+func NewTwoPCWorkerWrapper(runner *TwoPCRunner, serverID ServerID) *TwoPCWorkerWrapper {
+	return &TwoPCWorkerWrapper{
+		serverID: serverID,
+		runner:   runner,
+	}
+}
+
+// Prepare implements twopc.Worker.Prepare
+func (tpww *TwoPCWorkerWrapper) Prepare(ctx context.Context, wb twopc.WriteBatch) error {
+	return tpww.callRemote("Prepare", ctx, wb)
+}
+
+// Commit implements twopc.Worker.Commit
+func (tpww *TwoPCWorkerWrapper) Commit(ctx context.Context, wb twopc.WriteBatch) error {
+	// extract log index only
+	l, ok := wb.(*Log)
+	if !ok {
+		return ErrInvalidLog
+	}
+
+	return tpww.callRemote("Commit", ctx, l.Index)
+}
+
+// Rollback implements twopc.Worker.Rollback
+func (tpww *TwoPCWorkerWrapper) Rollback(ctx context.Context, wb twopc.WriteBatch) error {
+	// extract log index only
+	l, ok := wb.(*Log)
+	if !ok {
+		return ErrInvalidLog
+	}
+
+	return tpww.callRemote("Rollback", ctx, l.Index)
+}
+
+func (tpww *TwoPCWorkerWrapper) callRemote(method string, ctx context.Context, args interface{}) error {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
context.Context should be the first parameter of a function